[release/3.1] Expose all references when not restoring

dougbu opened the pull request:
- backport of 0b0bed3eb89a (#32718)

* Expose all references when not restoring
  - use empty `$(MSBuildRestoreSessionId)` to determine when contributing to dependency graph
* Remove extra direct references
  - should now be part of the dependency graph automatically
* Avoid errors about non-shared Fx references
  - not a problem unless executing `restore` target
* Special case source builds

dougbu committed:
[release/3.1] Expose all references when not restoring
- backport of 0b0bed3eb89a (#32718)

* Expose all references when not restoring
  - use empty `$(MSBuildRestoreSessionId)` to determine when contributing to dependency graph
* Remove extra direct references
  - should now be part of the dependency graph automatically
* Avoid errors about non-shared Fx references
  - not a problem unless executing `restore` target
* Special case source builds

wtgodbe commented:
Looks like this is causing SharedFx tests to fail

`SharedFrameworkContainsListedAssemblies`

> Assert.Empty() Failure\nCollection: ["System.Text.Json", "Microsoft.Bcl.AsyncInterfaces"]

>    at Microsoft.AspNetCore.SharedFxTests.SharedFrameworkContainsListedAssemblies() in /_/src/Framework/test/SharedFxTests.cs:line 50

dougbu committed:
Add workaround for ref/ generation
- use `$(CompileUsingReferenceAssemblies)` to avoid new references in generated projects

dougbu committed:
Limit `_CheckForReferenceBoundaries` to Core TFM
- back-ports a very small part of 5266918ed2b
- avoid incorrect errors about extra references in shared framework
- safe because every `$(IsAspNetCoreApp)` project targets (or multi-targets) `netcoreapp3.1`

dougbu commented:
This is blocked for now because the Arcade SDK doesn't set `$(MSBuildRestoreSessionId)` when restoring the repo. I tried back-porting dotnet/arcade@c2875f0932d but that caused new issues related (I think) to broader use of Helix.targets in the release/3.x branch. I tried back-porting dotnet/arcade@961b7d2e1c7 as well to resolve those new problems but, while it minimized the differences `git diff release/5.0 -- src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj` shows (mostly to publish bits), the result was lots of new warnings in my local builds of dotnet/aspnetcore.

/btw dotnet/arcade@c2875f0932d alone doesn't work in release/3.x because that uses a `<PropertyGroup>` to create `@(_SolutionRestoreProps)` items. I needed to move that into an `<ItemGroup>`, perhaps due to the different .NET SDK. (release/5.0 builds are fine and we use desktop `msbuild` at least some of the time in both branches.)

Any suggestions @rainersigwald @ViktorHofer❔

dougbu committed:
Add more `$(MSBuildRestoreSessionId)` checks
- additional projects in this branch have TFM-specific references

dougbu committed:
nit: Remove a few duplicate references
- no need to mention Microsoft.AspNetCore.DataProtection twice

dougbu commented:
The current changes do **not** remove all CS1705 workarounds. Still need a lot of `<CompileUsingReferenceAssemblies>false</CompileUsingReferenceAssemblies>` in our projects to avoid compiling against our ref/ assemblies. These workarounds are specific to this branch and help e.g. with transitive references through EF assemblies.

dougbu commented:
> Still need a lot of `<CompileUsingReferenceAssemblies>false</CompileUsingReferenceAssemblies>` in our projects

Good news is these workarounds are all in test (or test asset), sample, and (in one case) benchmark projects.

dougbu commented:
Temporarily blocked on dotnet/arcade#7612 and an Arcade dependency update in our 'release/3.1' after that. I'll rebase and remove the extra commits to use 'General Testing' assemblies when that'll work&hellip;

dougbu commented:
Please review but ignore 92376e7dfef89f5d8ce5b2e5274f7eb4c9ec7271 changes

wtgodbe reviewed (approved):
LGTM

dougbu committed:
Remove excess assemblies from shared Fx
- brought in due to improved reference discovery

dougbu commented:
This is no longer `blocked` and is ready for `3.1.20`.

wtgodbe commented:
My memory is foggy now, what problem was this trying to solve?

dougbu commented:
> what problem was this trying to solve?

Remember our STEW and STJ issues in 5.0.x❔ That problem stemmed from _not_ finding transitive references and thereby missing unexpected assembly versions in our tests. This change ensures all assemblies from NETCore.App and the transport package (in later branches) are exposed, ship correctly, and verified in our tests.
